xds/internal/test/e2e/e2e_test.go

```diff
@@ -0,0 +1,262 @@
+/*
+ *
+ * Copyright 2021 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package e2e
+
+import (
+	"bytes"
+	"context"
+	"flag"
+	"fmt"
+	"os"
+	"strconv"
+	"testing"
+	"time"
+
+	v3clusterpb "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
+	v3routepb "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
+	channelzpb "google.golang.org/grpc/channelz/grpc_channelz_v1"
+	testpb "google.golang.org/grpc/interop/grpc_testing"
+	"google.golang.org/grpc/xds/internal/testutils/e2e"
+)
+
+var (
+	clientPath = flag.String("client", "./binaries/client", "The interop client")
+	serverPath = flag.String("server", "./binaries/server", "The interop server")
+)
+
+func TestMain(m *testing.M) {
+	flag.Parse()
+	if _, err := os.Stat(*clientPath); os.IsNotExist(err) {
+		return
+	}
+	if _, err := os.Stat(*serverPath); os.IsNotExist(err) {
+		return
+	}
+	os.Exit(m.Run())
+}
+
+type testOpts struct {
+	testName     string
+	backendCount int
+	clientFlags  []string
+}
+
+func setup(t *testing.T, opts testOpts) (*controlPlane, *client, []*server) {
+	t.Helper()
+	backendCount := 1
+	if opts.backendCount != 0 {
+		backendCount = opts.backendCount
+	}
+
+	cp, err := newControlPlane(opts.testName)
+	if err != nil {
+		t.Fatalf("failed to start control-plane: %v", err)
+	}
+	t.Cleanup(cp.stop)
+
+	var clientLog bytes.Buffer
+	c, err := newClient(fmt.Sprintf("xds:///%s", opts.testName), *clientPath, cp.bootstrapContent, &clientLog, opts.clientFlags...)
+	if err != nil {
+		t.Fatalf("failed to start client: %v", err)
+	}
+	t.Cleanup(c.stop)
+
+	var serverLog bytes.Buffer
+	servers, err := newServers(opts.testName, *serverPath, cp.bootstrapContent, &serverLog, backendCount)
+	if err != nil {
+		t.Fatalf("failed to start server: %v", err)
+	}
+	t.Cleanup(func() {
```

Comment:
Should we make the `newControlPlane`, `newClient`, and `newServer` functions take a `testing.T` and have them call `t.Cleanup()` instead?

Reply:
I didn't want to make them only helper functions for tests.
They can be used by binaries, maybe help with developing (e.g. a binary that starts the control plane and backends, so we can iterate on the client).

For the same reason I kept them in `e2e.go`, not `_test.go`.
But I don't have a concrete idea how that works...

Reply:
Ack. 
We might have to move it a different package, something like `grpcutil` or `xdsutil` at some point after we move out all tests from `xds/internal/test` to `grpc/test`. But that shouldn't affect this PR.
LGTM.